src/core/components/pubsub.js

```diff
@@ -16,65 +16,56 @@ module.exports = function pubsub (self) {
         options = {}
       }
 
-      if (!self._options.EXPERIMENTAL.pubsub) {
+      if (!self.libp2p.pubsub) {
         return callback
           ? setImmediate(() => callback(errPubsubDisabled()))
           : Promise.reject(errPubsubDisabled())
       }
 
       if (!callback) {
-        return new Promise((resolve, reject) => {
-          self.libp2p.pubsub.subscribe(topic, options, handler, (err) => {
-            if (err) {
-              return reject(err)
-            }
-            resolve()
-          })
-        })
+        return self.libp2p.pubsub.subscribe(topic, handler, options)
       }
 
-      self.libp2p.pubsub.subscribe(topic, options, handler, callback)
+      self.libp2p.pubsub.subscribe(topic, handler, options, callback)
     },
 
     unsubscribe: (topic, handler, callback) => {
-      if (!self._options.EXPERIMENTAL.pubsub) {
+      if (!self.libp2p.pubsub) {
         return callback
           ? setImmediate(() => callback(errPubsubDisabled()))
           : Promise.reject(errPubsubDisabled())
       }
 
-      self.libp2p.pubsub.unsubscribe(topic, handler)
-
       if (!callback) {
-        return Promise.resolve()
+        return self.libp2p.pubsub.unsubscribe(topic, handler)
       }
 
-      setImmediate(() => callback())
+      self.libp2p.pubsub.unsubscribe(topic, handler, callback)
```

Comment:
Can this be simplified? Instead of:

```js
      if (!callback) {
        return self.libp2p.pubsub.unsubscribe(topic, handler)
      }

      self.libp2p.pubsub.unsubscribe(topic, handler, callback)
```

Just:

```js
      return self.libp2p.pubsub.unsubscribe(topic, handler, callback)
```

As in, shouldn't promisify deal with a `null`/`undefined` `callback` and return a `Promise`?

Reply:
Let me call @jacobheun here. 

He added this, but I think there was an issue here with `promisify`

Reply:
Promisify can't handle this because it adds extra args to the call. Promisify will see that the last argument is not a function, in this instance it is `null` or `undefined`. Promisify will then add a callback function and call unsubscribe. So you end up with the args `[topic, handler, undefined, promisifyCallback]`, and unsubscribe will internally call `undefined` thinking it's the callback.